src/InsertionsClient.Console/ConsoleApp/BuildFilterFactory.cs

```diff
@@ -0,0 +1,246 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using Microsoft.DotNet.InsertionsClient.Models;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq.Expressions;
+using System.Text;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.DotNet.InsertionsClient.ConsoleApp
+{
+    internal static class BuildFilterFactory
+    {
+        /// <summary>
+        /// Given a filter string, generates a predicate that will decide
+        /// if the given build should pass or fail the filter.
+        /// </summary>
+        /// <param name="filterString">Input string that defines the filter.</param>
+        /// <param name="buildFilter">The created predicate.</param>
+        /// <returns>True if operation succeeded. False, otherwise.</returns>
+        public static bool TryCreateFromString(string filterString, out Predicate<Build>? buildFilter)
+        {
+            buildFilter = null;
+
+            if (string.IsNullOrWhiteSpace(filterString))
+            {
+                Trace.WriteLine("Failed to parse build filter. Provided input string is empty.");
+                return false;
+            }
+
+            try
+            {
+                // Root expression that the final predicate will be build on.
+                // Generated expressions from each ruleset will be combined with "Or" to construct this root expression.
+                Expression? rootExpression = null;
+
+                // Parameter for the build that we pass to the filter i.e. the parameter for the final predicate.
+                ParameterExpression buildParameter = Expression.Parameter(typeof(Build), "build");
```

Comment:
consider caching this `typeof()` call as a static field

Reply:
I believe this falls a bit into "premature optimization" (don't mind the overused, ugly term here) considering this isn't a hot-path, but I squeezed it into the PR.

Reply:
Looks like I accidentally merged before I could push all the changes. I will create a new PR for this.

Reply:
Any time you're executing typeof() for a non-variable Type in callable methods, you should cache the value.  
- How do you distinguish premature optimizations from best practices?
- we should default in favor of best practice & cheaper execution, unless other variables, more important in that situation, demand doing otherwise; such as code maintenance.

Reply:
I agree that we should default in favor of best practices. For this specific case, I think we are hurting maintainability/readability more than we gain in performance (which, in my opinion is quite negligible here).